python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -248,18 +223,18 @@ def create_node(
             url = "rayclusters/{}".format(self.cluster_name)
             raycluster = self._get(url)
             group_name = tags["ray-user-node-type"]
-            group_index, group_spec = self._get_worker_group(raycluster, group_name)
+            group_index, _ = self._get_worker_group(raycluster, group_name)
+            tag_filters = {TAG_RAY_USER_NODE_TYPE: groupName}
+            current_replica_count = len(self.non_terminated_nodes(tag_filters))
```

Comment:
just fyi: The fact that we need to call `non_terminated_nodes` here is slightly unfortunate -- ideally we would only call `non_terminated_nodes` once per autoscaler iteration and there was probably a place earlier in the autoscaler that was reading non_terminated nodes and as a result decided that scaling up was the right thing to do and maybe we could re-use the non_terminated_nodes from that. For code simplicity it is probably better to call it again rather than try to thread it through, so let's go with what you have -- overall this is a really great improvement to what we had earlier and there is no need to try to make it perfect :)

Reply:
Let's chat about this tomorrow - we can figure out how to refactor this. We are also doing something similar in terminate_nodes.

Reply:
Hmm, yep, same thing in terminate_nodes. An issue is that a `list` without specifying a resource version is bad for scalability. But in practice it's not that many more calls here.

There are some environment variables we can set to cap the potential number of "create nodes" per autoscaler iteration to the number of worker groups.

Reply:
It would also be possible to cache the results of the autoscaler's `non_terminated_nodes({})` call. 

Then create_node and terminate_node would do some modification of the cached values,
which is more or less exactly what the autoscaler already does in the `_update` method.

This would be safe since everything is behind self._lock.

I think Philipp was suggesting somehow reusing the autoscaler's cacheing work, but that would be hard.

Anyways, unless there's a simpler approach, not clear if we want to refactor in this PR.